### ISSUE TITLE
Towards true non-flakiness

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis#534]
In #527, @bukzor quite correctly points out that the claim that Hypothesis currently has an invariant of non-flakiness is false: There are some things a user can do that will result in flaky tests. These were added before I got quite so fanatical about the subject, and have since been grandfathered in.

The solution to this is to make Hypothesis *actually* non-flaky. This is a holding bug for planning how to do that.

The following are the cases I am currently aware of:

* [ ] when assume() or filter() or are hard to satisfy, this can result in Hypothesis failing to reach min\_satisfying\_examples with non-zero probability.
* [ ] timeout makes the number of tests you run dependent on system load etc which again can cause the test to fail because of min\_satisfying\_examples.

Can anyone think of any others?

Any solutions to these problems will require some rather tricky deprecation paths I think.

Rather than propose solutions in this issue text I'll add comments about them.

**Comment by alexwlchan** [comment]
Did you get any further with #523?

**Comment by DRMacIver** [comment]
I didn't -weirdly I haven't seen it fail since the change - but note that this is about non-flakiness for Hypothesis *users*. We're stuck with flakiness more or less intrinsically.

**Comment by DRMacIver** [comment]
Here are my current somewhat half-baked thoughts on the subject:

* I think timeout was a bad feature and we should remove it. It's been the source of a lot of grief, the default is stupidly high anyway, and I feel like it's more or less intrinsically problematic. The deprecation path here will be somewhat tricky - I think we should start warning users who are hitting the timeout and suggest they set max\_examples.
* I think we should deprecate min\_satisfying\_examples altogether and replace it with a better heuristic. I'm not sure what yet.
* Whatever heuristic we use, the engine should run either run fully deterministically until it is satisfied, or we should be able to deterministically determine whether it would have been satisfied independently of what we actually ran.
* A very simple way of achieving the above is to use the derandomize seed until we satisfy the condition and then switch to the provided Random instance once we have. I don't especially like this solution, but it is a valid one and probably a decent first pass.

**Comment by Zac-HD** [comment]
`test_interleaving_engines` is still flaky - [example build](https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/224772882).  I don't know if #511 missed something, or maybe it's really failing?

**Comment by DRMacIver** [comment]
Still not an issue about our build being flaky. :-) I've opened another ticket about that though. https://github.com/HypothesisWorks/hypothesis-python/issues/571

**Comment by bukzor** [comment]
@DRMacIver These have changed significantly (once #1211 is merged). Are the problems you mentioned gone? Have new ones appeared?

> when assume() or filter() or are hard to satisfy, this can result in Hypothesis failing to reach min_satisfying_examples with non-zero probability.
> timeout makes the number of tests you run dependent on system load etc which again can cause the test to fail because of min_satisfying_examples.
```
